### PR TITLE
ARTEMIS-2024 Enable SharedClientID on ConnectionFactory

### PR DESCRIPTION
**Pull request by mtaylor**

### CODE REVIEW

File: artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQConnection.java [hunk 1]
```diff
@@ -673,12 +673,16 @@ protected final void checkClosed() throws JMSException {
       }
    }
 
-   public void authorize() throws JMSException {
+   public void authorize(boolean validateClientId) throws JMSException {
```
Comment (michaelandrepearce) [inline artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQConnection.java]: public interface change in client code, should probably keep the existing method for people who may have coded against it.

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/jms/client/ConnectionTest.java [hunk 2]
```diff
@@ -131,6 +131,27 @@ public void testTwoConnectionsSameIDThroughCF() throws Exception {
...
+
+      conn = connectionFactory.createConnection();
+      try {
+         conn2 = connectionFactory.createConnection();
+      } catch (InvalidClientIDException expected) {
+         Assert.fail("Should allow sharing of client IDs among the same CF");
+      }
+
+      Session session1 = conn.createSession();
+      Session session2 = conn.createSession();
```
Comment (michaelandrepearce) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/jms/client/ConnectionTest.java]: Probably should test conn2 as well here?

File: artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQConnection.java (hunk 1 above)
Comment (clebertsuconic) [reply artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQConnection.java]: @michaelandrepearce that's internal API though

**Comment by clebertsuconic** [comment]
failed test:
Failed tests: 
  ConnectionFactoryPropertiesTest.testCompareConnectionFactoryAndResourceAdapterProperties:86->compare:97->Assert.fail:88 in ActiveMQ Connection Factory only: [enableSharedClientID]
in ActiveMQ Resource Adapter only: []

File: artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQConnection.java (hunk 1 above)
Comment (mtaylor) [reply artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQConnection.java]: The method was public I will add it back.

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/jms/client/ConnectionTest.java (hunk 2 above)
Comment (mtaylor) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/jms/client/ConnectionTest.java]: Yes.  Copy and paste error ;).  Will fix.

**Comment by clebertsuconic** [comment]
@jmesnil TBH: I thinks it's dumb to use clientID on a standalone CF.

**Comment by michaelandrepearce** [comment]
You have to set it for non shared durable subscriber by Jms spec.

**Comment by clebertsuconic** [comment]
@michaelandrepearce on the connection.. yeah! On the connectionFactory. .that's a broker-specific extension. you set it on the CF as an utility for the user.

That's just that! An utility. Setting it on the CF or on the connection it should have the same effect.

**Comment by mtaylor** [comment]
The spec also says that the preferred way to set the client id is via the CF.  I think we just have to chalk this one up as the spec being ambiguous.  I can see how people may interpret one way or the other.

**Pull request merged**